python/instrumentation/openinference-instrumentation-vertexai/src/openinference/instrumentation/vertexai/_wrapper.py

```diff
@@ -0,0 +1,469 @@
+UsageMetadata: TypeAlias = Union[
+    v1.GenerateContentResponse.UsageMetadata,
+    v1beta1.GenerateContentResponse.UsageMetadata,
+]
+GenerateContentRequest: TypeAlias = Union[
+    v1.GenerateContentRequest,
+    v1beta1.GenerateContentRequest,
+]
+GenerateContentResponse: TypeAlias = Union[
+    v1.GenerateContentResponse,
+    v1beta1.GenerateContentResponse,
+]
+
+_PREDICTION_SERVICE_REQUESTS = sorted(
+    {
+        name
+        for module in (
+            # https://cloud.google.com/vertex-ai/docs/reference#versions
+            # - v1: Stable
+            # - v1beta1: Supports latest preview features
+            v1.types.prediction_service,
+            v1beta1.types.prediction_service,
+        )
+        for name, _ in inspect.getmembers(
+            module,
+            lambda cls: inspect.isclass(cls)
+            and issubclass(cls, proto.Message)
+            and cls.__name__.endswith("Request"),
+        )
+    }
+)
+
+
+class _Wrapper:
+    _status = _status
+
+    def __init__(self, tracer: trace_api.Tracer, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+        self._tracer = tracer
+
+    @wrapt.decorator  # type: ignore[misc]
+    def __call__(
+        self,
+        wrapped: Callable[..., Any],
+        instance: Any,
+        args: Tuple[type, Any],
+        kwargs: Mapping[str, Any],
+    ) -> Any:
+        if (
+            not (request := _get_prediction_service_request(chain(args, kwargs.values())))
+            or context_api.get_value(_SUPPRESS_INSTRUMENTATION_KEY)
+            or not self._status._IS_INSTRUMENTED
+        ):
+            return wrapped(*args, **kwargs)
+        span_name = inspect.stack()[1].function
+        attributes = dict(get_attributes_from_context())
+        span = self._tracer.start_span(name=span_name, attributes=attributes)
+        span.set_attribute(OPENINFERENCE_SPAN_KIND, LLM)
+        if isinstance(request, proto.Message):
+            span.set_attribute(INPUT_VALUE, safe_json_dumps(request.__class__.to_dict(request)))
+            span.set_attribute(INPUT_MIME_TYPE, JSON)
+        else:
+            span.set_attribute(INPUT_VALUE, str(request))
+        try:
+            _update_span(request, span)
+        except BaseException:
+            pass
+        try:
+            with _use_span(span)():
+                result = wrapped(*args, **kwargs)
+        except BaseException as exc:
+            span.record_exception(exc)
+            span.set_status(Status(StatusCode.ERROR, f"{type(exc).__name__}: {exc}"))
+            span.end()
+            raise
+        if isinstance(result, Awaitable):
+            return _proxy(result, _CallbackForAwaitable(request, span), _use_span(span))
+        if isinstance(result, (Iterable, AsyncIterable)):
+            return _proxy(result, _CallbackForIterable(request, span), _use_span(span))
+        _finish(span, result)
+        return result
+
+
+class _CallbackForAwaitable:
```

Comment:
some docstrings here could help a first time reader understand the unitily or need to have these. and how to use